migration-engine/ARCHITECTURE.md

```diff
@@ -117,6 +118,41 @@ On a high level, migrate deploy _exclusively_:
   looking at the `_prisma_migrations` table in the target database.
 - Run the migrations that have not been applied yet, in chronological order.
 
+`prisma migrate deploy` is the command meant to be used for _unattended_
+migrations (as part of CI). As such, it should be as reliable, predictable and
+deterministic as possible.
+
+### Why does `prisma migrate deploy` not warn when a migration is in the migrations table but not in the migrations folder?
+
+We don't want to warn if an already applied migration is missing from the
+migration directory because it would prevent squashing migrations and
+deployment from not-properly-rebased branches.
+
+In general, by design, `deploy` errs on the side of not standing in the way of
+deploying your migrations.
+
+### Why does `prisma migrate deploy` not detect drift?
+
+We don't detect drift because we want to keep the deployment path light and
+simple, and because we would need a shadow database for that. Many people would
+not be comfortable with creating/using temporary databases being on the
+deployment path. And as stated in the previous question, in general, by design,
+`deploy` errs on the side of not standing in the way of deploying your
+migrations.
+
+### What happens when a migration fails in `prisma migrate deploy`?
+
+If a migration fails during deployment, you will see the error. Then `prisma
+migrate status` and `prisma migrate deploy` commands (in subsequent runs) will
+show you the failed state with the error message when you run them again.
+
+It's then your responsibility to fix what failed, and marked the migration as
+applied or rolled back with `prisma migrate resolve`, so you can deploy
+migrations again.
+
+Prisma Migrate does not offer much help at that last stage, but this is
+something we are working on defining and prioritizing.
```

Comment:
Issue link?

Reply:
We had issues that were closed. I'll open a new placeholder one.

Reply:
Added a link to the new issue and pushed.